cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -592,6 +592,39 @@ VersionedItem LocalVersionedEngine::write_versioned_metadata_internal(
     }
 }
 
+std::vector<VersionedItem> LocalVersionedEngine::batch_write_versioned_metadata_internal(
+    std::vector<StreamId> stream_ids,
+    bool prune_previous_versions,
+    std::vector<arcticdb::proto::descriptors::UserDefinedMetadata>&& user_meta_proto_vec) {
+    auto stream_update_info_vector_fut = batch_get_latest_undeleted_version_and_next_version_id_async(store(),
+                                                                                                      version_map(),
+                                                                                                      stream_ids);
+    std::vector<folly::Future<VersionedItem>> output_vect_fut;
+    for(const auto& stream_id : folly::enumerate(stream_ids)) {
+        output_vect_fut.push_back(
+            std::move(stream_update_info_vector_fut[stream_id.index])
+            .thenValue([this, &stream_id, user_meta_proto = std::move(user_meta_proto_vec[stream_id.index])](auto&& update_info) mutable {
+                util::check(update_info.previous_index_key_.has_value(),
+                        "Failed to find latest version to write metadata to for symbol {}",
+                        *stream_id);
+                return async::submit_io_task(UpdateMetadataTask{store(), update_info, std::move(user_meta_proto)}).thenValue([update_info = std::move(update_info)](auto&& index_key) mutable {
+                    return std::pair<AtomKey, version_store::UpdateInfo>(std::move(index_key), std::move(update_info));
+                });
+            })
+            .thenValue([this, prune_previous_versions](auto&& index_key_update_info_pair){
+                auto index_key = std::move(index_key_update_info_pair.first);
+                if(prune_previous_versions) {
+                    return async::submit_io_task(WriteAndPrunePreviousTask{store(), version_map(), index_key, index_key_update_info_pair.second.previous_index_key_})
+                    .then([index_key = std::move(index_key)](auto&& ){return index_key;});
+                } else {
+                    return async::submit_io_task(WriteVersionTask{store(), version_map(), index_key})
+                    .then([index_key = std::move(index_key)](auto&& ){return index_key;});
```

Comment:
This duplicated code can be in a single `thenValue` chained to this one

Reply:
I don’t understand your comment here. I cannot capture the value in the next lambda because update_info is the value wrappered inside the future, which also needs to be propagated to the next future. On the other side, I need the ThenValue because one of the elements still is a future (what submit_io_task returns) and the other is an actual value, so I need to reach the point in time where both of them are actual values.

Reply:
agree